### DISCUSSION TITLE
Terminal UI - No pane when running single task

### DISCUSSION THREAD

**Discussion by un33k** [vercel/turborepo#8351] [Ideas]
### Goals

1. Projects that don't need tasks.
2. Ability to copy and paste from the Task window (not possible now).
3. Ability to turn tasks off and see everything like before.

### Non-goals

1.
2.
3.

### Background

The new task window is extremely hard to work with, not easy to copy out of the windows. The lack of color makes it hard to find things, and the continuous spinners next to `watch`ed tasks are reducing the overall quality of the experience using Turbo.

"turbo": "^2.0.2-canary.1"

The task windows added value is negative at this time (objective).
The quality of development experience has gone down drastically ( subjective).

### Proposal

It would be great to opt out of the visual representation of 2.0 task windows.
Something like adding `taskWindows: false` to the `turbo.json`.

**Comment by un33k** [comment]
**When we run `next dev`:**

<img width="989" alt="Screenshot 2024-06-06 at 8 59 56 AM" src="https://github.com/vercel/turbo/assets/407537/fdd8c144-a3a0-4d44-befc-5f6d3a910bac">

.
.
.


**Pressed `Enter` plus 20 `arrow down` later to get to the actual apps running:**

<img width="990" alt="Screenshot 2024-06-06 at 9 00 14 AM" src="https://github.com/vercel/turbo/assets/407537/ea2a997b-66d7-492f-9a46-7cad1619710d">

.
.
.


**No visual indication of a failed library, unless we navigate to it. Spinner is running, without any indication:**

<img width="997" alt="Screenshot 2024-06-06 at 9 19 06 AM" src="https://github.com/vercel/turbo/assets/407537/0f46b2d3-9b20-45bc-8d90-43708d341dbc">

**Comment by anthonyshew** [comment]
If you prefer the previous UI with streamed logs, you're able to opt out of it using `ui` in `turbo.json`: https://turbo.build/repo/docs/reference/configuration#ui

A few things at a glance:
- The paned view is an objective improvement from the previous streamed logs UI. It lets you interact with tasks and more clearly see what is happening with each task individually. Additionally, it gives us the ability to more tightly control CLI tools that have a bad habit of wiping the entire terminal clean, which we received a lot of feedback about and have now solved with the new terminal UI.
- Colors being missing in your tasks is a bug. Colors should be the default. If you could open an issue for us on that (if there isn't one already), that would be great.
- We've heard other feedback that the icons could be 1) more informative as far about errored states and 2) less distracting by being less spinny. We'll work on this.

Definitely hear your feedback. There are papercuts to work out and we'll be working towards an even better experience here.

**Comment by MickL** [comment]
When starting just one task (e.g. with --filter="...") there is no need for the left sidebar. In this case the whole terminal UI can be deactivated.

**Comment by WilliamABradley** [comment]
To speed up developer experience, it would be nice to change the TUI view if only one task was selected via a filter.

This could either be automatic if there is only one task running, or via a flag, e.g. `--enter` or `--single`, that could involve doing the following:

1. Hide the Task sidebar (There is only one task)
2. Enter interact mode automatically.

Before turbo 2, we had a helper script for running local versions of our packages:

```ts
import { execSync } from "child_process";

const [, , packageName] = process.argv;
console.log(`Running local script for ${packageName}`);

// Unfortunately turbo doesn't support stdin forwarding, so this script does that heavy lifting.
const command = `turbo build --filter *${packageName} -- --env local && pnpm --filter *${packageName} local:exec`;
console.log(`Running command: ${command}`);
execSync(command, { stdio: "inherit" });
```

Now we can just do `turbo local --filter *my-package`, which we generally want to run 1 task for, however, to interact with it, we have to remember to press enter, which is a slight bit of inefficiency, and it is a little wasteful to show the sidebar.

I guess if stream mode supported stdin forwarding that would solve most of that, but that would fall apart with more than one task selected.
